harmony-kotlin/src/commonMain/kotlin/com.harmony.kotlin/data/datasource/RetryDataSource.kt

```diff
@@ -0,0 +1,72 @@
+package com.harmony.kotlin.data.datasource
+
+import com.harmony.kotlin.common.logger.Logger
+import com.harmony.kotlin.data.query.Query
+
+class RetryDataSource<V>(
+  private val getDataSource: GetDataSource<V>? = null,
+  private val putDataSource: PutDataSource<V>? = null,
+  private val deleteDataSource: DeleteDataSource? = null,
+  private val maxAmountOfRetries: Int = 2,
+  private val retryIf: (Exception) -> Boolean = { true },
+  private val logger: Logger? = null,
+) : GetDataSource<V>, PutDataSource<V>, DeleteDataSource {
+
+  init {
+    check(maxAmountOfRetries > 0)
+  }
+
+  override suspend fun get(query: Query): V =
+    executeWithRetries(retryCount = maxAmountOfRetries) {
+      getDataSource().get(query)
+    }
+
+  override suspend fun getAll(query: Query): List<V> =
+    executeWithRetries(retryCount = maxAmountOfRetries) {
+      getDataSource().getAll(query)
+    }
+
+  override suspend fun put(query: Query, value: V?): V =
+    executeWithRetries(retryCount = maxAmountOfRetries) {
+      putDataSource().put(query, value)
+    }
+
+  override suspend fun putAll(query: Query, value: List<V>?): List<V> =
+    executeWithRetries(retryCount = maxAmountOfRetries) {
+      putDataSource().putAll(query, value)
+    }
+
+  override suspend fun delete(query: Query) =
+    executeWithRetries(retryCount = maxAmountOfRetries) {
+      deleteDataSource().delete(query)
+    }
+
+  private suspend fun <K> executeWithRetries(retryCount: Int, functionToRetry: suspend () -> K): K =
+    @Suppress("TooGenericExceptionCaught")
+    try {
+      functionToRetry()
+    } catch (error: Exception) {
+      logger?.e("RetryDataSource", "failed attempt ${this.maxAmountOfRetries - retryCount + 1} of max retries ${this.maxAmountOfRetries}:\n\t$error")
```

Comment:
not sure on this... I would say that any failure here should be logged as an unexpected error, since we are just catching anything just to retry. In any case, changed.